src/fragments/lib/graphqlapi/js/optimistic-ui.mdx

```diff
@@ -0,0 +1,539 @@
+      >({
+        query: queries.listRealEstateProperties,
+      });
+
+      const allRealEstateProperties =
+        response?.data?.listRealEstateProperties?.items;
+
+      if (!allRealEstateProperties) return null;
+
+      return allRealEstateProperties;
+    },
+  });
+
+  // TanStack create mutation with optimistic updates
+  const createMutation = useMutation({
+    mutationFn: async (
+      realEstatePropertyDetails: CreateRealEstatePropertyInput
+    ) => {
+      const newRealEstateProperty = await API.graphql<
+        GraphQLQuery<CreateRealEstatePropertyMutation>
+      >({
+        query: mutations.createRealEstateProperty,
+        variables: { input: realEstatePropertyDetails },
+      });
+
+      console.log(newRealEstateProperty);
+      return newRealEstateProperty;
+    },
+    // When mutate is called:
+    onMutate: async (newRealEstateProperty) => {
+      // Cancel any outgoing refetches
+      // (so they don't overwrite our optimistic update)
+      await queryClient.cancelQueries({ queryKey: ["realEstateProperties"] });
+
+      // Snapshot the previous value
+      const previousRealEstateProperties = queryClient.getQueryData([
+        "realEstateProperties",
+      ]);
+
+      // Optimistically update to the new value
+      // TODO: fix type
+      // @ts-ignore
+      queryClient.setQueryData(["realEstateProperties"], (old) => [
+        // TODO: fix type
+        // @ts-ignore
+        ...old,
+        newRealEstateProperty,
+      ]);
+
+      // Return a context object with the snapshotted value
+      return { previousRealEstateProperties };
+    },
+    // If the mutation fails,
+    // use the context returned from onMutate to roll back
+    onError: (err, newRealEstateProperty, context) => {
+      queryClient.setQueryData(
+        ["realEstateProperties"],
+        // TODO: fix type
+        // @ts-ignore
+        context.previousRealEstateProperties
+      );
+    },
+    // Always refetch after error or success:
+    onSettled: () => {
+      queryClient.invalidateQueries({ queryKey: ["realEstateProperties"] });
+    },
+  });
+
+  function RealEstatePropertyDetailView() {
+    const {
+      data: realEstateProperty,
+      isLoading,
+      isSuccess,
+      isError: isErrorQuery,
+    } = useQuery({
+      queryKey: ["realEstateProperties", currentRealEstatePropertyId],
+      // TODO: improve return type
+      queryFn: async (): Promise<RealEstateProperty | null | undefined> => {
+        const response = await API.graphql<
+          GraphQLQuery<GetRealEstatePropertyQuery>
+        >({
+          query: queries.getRealEstateProperty,
+          variables: { id: currentRealEstatePropertyId },
+        });
+
+        console.log("RealEstatePropertyDetailView queryFn response", response);
+        return response.data?.getRealEstateProperty;
+      },
+    });
+
+    // TanStack update mutation with optimistic updates
+    const updateMutation = useMutation({
```

Comment:
something to verify with TanStack's samples but maybe it can benefit from more use-case-descriptive variable name like "updateRealEstateProperty"